packages/language/src/language-server/document-symbol-builder.ts

```diff
@@ -0,0 +1,326 @@
+    children: DocumentSymbol[],
+    textDocument: TextDocument,
+  ): LevelSymbol[] {
+    const levelSymbols: LevelSymbol[] = [];
+
+    let kind: SymbolKind = SymbolKind.Variable;
+    if (
+      // Type is lower case in the ast but upper case in the definition.
+      item.attributes.some(
+        (a) =>
+          a.kind === SyntaxKind.ComputationDataAttribute &&
+          a.type?.toString() === "constant",
+      )
+    ) {
+      kind = SymbolKind.Constant;
+    }
+
+    for (const element of item.elements.filter((e) => e !== "*")) {
+      if (element.kind === SyntaxKind.DeclaredItem) {
+        // Declared struct items can inherit their level from the parent,
+        // e.g., 11 (XX, ZZ) char(10).
+        levelSymbols.push(
+          ...this.retrieveLevelSymbols(
+            element,
+            item.level ?? inheritedLevel,
+            children,
+            textDocument,
+          ),
+        );
+      } else if (
+        element.kind === SyntaxKind.DeclaredVariable &&
+        element.nameToken &&
+        isValidToken(element.nameToken)
+      ) {
+        const name = element.name ?? "<unknown>";
+        const range = getSyntaxNodeRange(element) ?? { start: 0, end: 0 };
+        levelSymbols.push({
+          level: item.level ?? inheritedLevel,
+          symbol: createDocumentSymbol(
+            name,
+            kind,
+            range,
+            children,
+            textDocument,
+          ),
+        });
+      }
+    }
+
+    return levelSymbols;
+  }
+}
+
+class LevelHierarchyBuilder {
+  private levelMap: Map<number, DocumentSymbol>;
+  private rootSymbols: DocumentSymbol[];
+
+  constructor() {
+    this.levelMap = new Map();
+    this.rootSymbols = [];
+  }
+
+  addSymbol(level: number | null, symbol: DocumentSymbol): void {
+    if (!level) {
+      this.rootSymbols.push(symbol);
+      return;
+    }
+
+    let parentLevel = level - 1;
+    while (!this.levelMap.has(parentLevel) && parentLevel > 0) {
+      parentLevel--;
+    }
+
+    if (parentLevel > 0) {
+      const parentSymbol = this.levelMap.get(parentLevel);
+      if (parentSymbol) {
+        parentSymbol.children = [symbol, ...(parentSymbol.children ?? [])];
+        parentSymbol.kind = SymbolKind.Struct;
+        parentSymbol.range.end = getLastPositionOrDefault(
+          parentSymbol.children,
+          symbol.range.end,
+        );
+      }
+    } else {
+      this.rootSymbols.push(symbol);
+    }
+
+    this.levelMap.set(level, symbol);
+    // Remove all larger levels to reset position
+    for (const key of this.levelMap.keys()) {
+      if (key > level) {
+        this.levelMap.delete(key);
+      }
+    }
+  }
+
+  getRootSymbols(): DocumentSymbol[] {
+    return this.rootSymbols;
+  }
+}
+
+function isValidToken(token: IToken): boolean {
+  // Will fail for tokens with no end offset.
+  // E.g., when generated by the preprocessor.
+  return token.endOffset !== undefined && token.endOffset > 0;
+}
```

Comment:
Suggestion: We already have a `isValidToken` function in `tokens.ts` which does this check.